File: pkg/widget/checkbox.go
package widget

// Prefer lua/lux/checkbox.lua for new code (pure Lua, no Go dependency).
// This Go widget remains functional for backward compatibility and direct usage
// via lumina.Checkbox, but new applications should use the Lua version.

import "github.com/akzj/lumina/pkg/render"

// CheckboxState is the internal state for a Checkbox widget.
type CheckboxState struct {
	Checked bool
	Hovered bool
}

// Checkbox is the built-in Checkbox widget.
// Props: label (string), checked (bool, controlled), disabled (bool)
// onChange fires with the new checked value (bool) on click or Space keydown.
var Checkbox = &Widget{
	Name: "Checkbox",
	NewState: func() any {
		return &CheckboxState{}
	},
	Render: func(props map[string]any, state any) any {
		t := CurrentTheme
		s := state.(*CheckboxState)
		label, _ := props["label"].(string)
		disabled, _ := props["disabled"].(bool)

		// Controlled mode: if "checked" prop is explicitly set, use it
		checked := s.Checked
		if v, ok := props["checked"]; ok {
			if b, ok := v.(bool); ok {
				checked = b
			}
		}

		// Determine checkbox visual
		indicator := "[ ]"
		if checked {
			indicator = "[x]"
		}

		// Colors
		fg := t.Text
		indicatorFg := t.Primary
		if disabled {
			fg = t.Muted
			indicatorFg = t.Muted
		} else if s.Hovered {
			indicatorFg = t.Hover
		}

		indicatorNode := &render.Node{
			Type:    "text",
			Content: indicator,
			Style: render.Style{
				Foreground: indicatorFg,
				Right:      -1,
				Bottom:     -1,
			},
		}

		children := []*render.Node{indicatorNode}

		if label != "" {
			labelNode := &render.Node{
				Type:    "text",
				Content: " " + label,
				Style: render.Style{
					Foreground: fg,
					Right:      -1,
					Bottom:     -1,
				},
			}
			if disabled {
				labelNode.Style.Dim = true
			}
			children = append(children, labelNode)
		}

		boxNode := &render.Node{
			Type:      "hbox",
			Children:  children,
			Focusable: !disabled,
			Disabled:  disabled,
			Style: render.Style{
				Right:  -1,
				Bottom: -1,
			},
		}
		for _, ch := range children {
			ch.Parent = boxNode
		}

		return boxNode
	},
	OnEvent: func(props map[string]any, state any, event *render.WidgetEvent) bool {
		s := state.(*CheckboxState)
		disabled, _ := props["disabled"].(bool)
		if disabled {
			return false
		}

		switch event.Type {
		case "mouseenter":
			if !s.Hovered {
				s.Hovered = true
				return true
			}
		case "mouseleave":
			if s.Hovered {
				s.Hovered = false
				return true
			}
		case "click":
			// Sync internal state from controlled prop before toggling
			if v, ok := props["checked"]; ok {
				if b, ok := v.(bool); ok {
					s.Checked = b
				}
			}
			s.Checked = !s.Checked
			event.FireOnChange = s.Checked
			return true
		case "keydown":
			if event.Key != " " && event.Key != "Enter" {
				return false
			}
			// Sync internal state from controlled prop before toggling
			if v, ok := props["checked"]; ok {
				if b, ok := v.(bool); ok {
					s.Checked = b
				}
			}
			s.Checked = !s.Checked
			event.FireOnChange = s.Checked
			return true
		}
		return false
	},
}

File: lua/lux/checkbox.lua
-- lua/lux/checkbox.lua
-- Pure Lua Checkbox component for Lux (controlled pattern)
-- Usage: local Checkbox = require("lux.checkbox")

local Checkbox = lumina.defineComponent("LuxCheckbox", function(props)
    local t = lumina.getTheme and lumina.getTheme() or {}
    local checked = props.checked or false
    local disabled = props.disabled == true
    local label = props.label or ""
    local hovered, setHovered = lumina.useState("hover", false)

    -- Visual
    local indicator = checked and "[x]" or "[ ]"
    local fg = disabled and (t.muted or "#6C7086") or (t.text or "#CDD6F4")
    local indicatorFg = disabled and (t.muted or "#6C7086")
        or (hovered and (t.hover or "#B4BEFE") or (t.primary or "#89B4FA"))

    local children = {
        lumina.createElement("text", {
            key = "indicator",
            foreground = indicatorFg,
        }, indicator),
    }
    if label ~= "" then
        children[#children + 1] = lumina.createElement("text", {
            key = "label",
            foreground = fg,
            dim = disabled,
        }, " " .. label)
    end

    return lumina.createElement("hbox", {
        id = props.id,
        key = props.key,
        focusable = not disabled,
        disabled = disabled,
        style = props.style,
        onClick = not disabled and function()
            if props.onChange then
                props.onChange(not checked)
            end
        end or nil,
        onKeyDown = not disabled and function(e)
            local k = type(e) == "table" and (e.key or e.Key) or tostring(e or "")
            if k == " " or k == "Enter" then
                if props.onChange then
                    props.onChange(not checked)
                end
            end
        end or nil,
        onMouseEnter = not disabled and function() setHovered(true) end or nil,
        onMouseLeave = not disabled and function() setHovered(false) end or nil,
    }, table.unpack(children))
end)

return Checkbox
